src/libs/Violations/ViolationsUtils.ts

```diff
@@ -0,0 +1,195 @@
+                return translate('violations.smartscanFailed');
+            case 'taxAmountChanged':
+                return translate('violations.taxAmountChanged');
+            case 'receiptRequired':
+                return translate('violations.receiptRequired', {
+                    amount: violation.data?.amount ?? '0',
+                    category: violation.data?.category ?? '',
+                });
+            case 'cashExpenseWithNoReceipt':
+                return translate('violations.cashExpenseWithNoReceipt', {amount: violation.data?.amount ?? ''});
+            case 'categoryOutOfPolicy':
+                return translate('violations.categoryOutOfPolicy');
+            case 'modifiedDate':
+                return translate('violations.modifiedDate');
+            case 'taxRateChanged':
+                return translate('violations.taxRateChanged');
+            case 'customUnitOutOfPolicy':
+                return translate('violations.customUnitOutOfPolicy');
+            case 'overLimit':
+                return translate('violations.overLimit', {amount: violation.data?.amount ?? ''});
+            case 'overLimitAttendee':
+                return translate('violations.overLimitAttendee', {amount: violation.data?.amount ?? ''});
+            case 'receiptNotSmartScanned':
+                return translate('violations.receiptNotSmartScanned');
+            case 'allTagLevelsRequired':
+                return translate('violations.allTagLevelsRequired');
+            case 'nonExpensiworksExpense':
+                return translate('violations.nonExpensiworksExpense');
+            case 'overAutoApprovalLimit':
+                return translate('violations.overAutoApprovalLimit', {formattedLimitAmount: violation.data?.formattedLimitAmount ?? ''});
+            case 'duplicatedTransaction':
+                return translate('violations.duplicatedTransaction');
+            case 'someTagLevelsRequired':
+                return translate('violations.someTagLevelsRequired');
+            case 'overCategoryLimit':
+                return translate('violations.overCategoryLimit', {categoryLimit: violation.data?.categoryLimit ?? ''});
+            case 'perDayLimit':
+                return translate('violations.perDayLimit', {limit: violation.data?.limit ?? ''});
+            default:
+                // The interpreter should never get here because the switch cases should be exhaustive.
+                //
+                // If typescript is showing an error on the assertion below it means the switch statement is out of
+                // sync with the `ViolationNames` type, and one or the other needs to be updated.
+                //
+                // The 'unnecessary' type assertion below is the guarantee of type safety for these translations.
+                //
+                // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+                return violation.name as never;
```

Comment:
Would it be better to return empty string here?

Reply:
No because this is the guarantee of exhaustiveness.

We want it to be clear that if the string passed in doesn't not match  one of the other cases that something has gone wrong. (And we want that to be a compile time error, not a runtime error.)

I originally omitted the default case, but that also causes a linter error, and I think it's more dangerous overall.

I felt like this was a more important guard because if you weren't paying attention you might see the require-default linter error, and remove it and add a default case.

This would break the guarantee that you will get back the right translation key *and* an object of the right shape. But there would be no errors when you did it.

This is at least conspicuous enough to be noticed, and if someone does something wrong (I.e. Modifies the list of violations without adding a message to be displayed) they compiler will flag this line as an error and they'll come and find the comment and add an entry to the switch statement.

The other way someone would say "Oh this doesn't have a default case, I'll add one" and there would not be any errors at all for any future changes to the list of violations.

Reply:
🤩 thanks for the explanation, I love it, particularly this _And we want that to be a compile time error, not a runtime error_

Reply:
Yeah, it's the best thing about typescript. When your codebase is properly typed, then when you make a change you can just follow the type errors through the code and the compiler shows you what needs to change.


There's obviously a bit of pain as you transition -- like trying to type the translate params, but once it gets to a certain critical mass it makes things so much easier.